ql/src/java/org/apache/hadoop/hive/ql/optimizer/GenMapRedUtils.java

```diff
@@ -1309,6 +1312,22 @@ public static void createMRWorkForMergingFiles(FileSinkOperator fsInput,
     // 2. Constructing a conditional task consisting of a move task and a map reduce task
     //
     Path inputDirName = fsInputDesc.getMergeInputDirName();
+
+    // For external CTAS queries, there is an optimisation which avoids rename/copy of
+    // files from -tmp to -ext. In such cases, the input dir must be -tmp.
+    try {
+      if (fsInputDesc.getTable() != null) {
+        FileSystem fs = inputDirName.getFileSystem(conf);
+        if (fsInputDesc.isCTASorCM() && TableType.EXTERNAL_TABLE.equals(fsInputDesc.getTable().getTableType())
+            && BlobStorageUtils.isBlobStorageFileSystem(conf, fs)
+            && !Utilities.shouldAvoidRename(fsInputDesc, conf)) {
+          fsInputDesc.setChangeToTmpPath(true);
```

Comment:
why not simply `fsInputDesc.changeToTmpPath()`?

Reply:
Done.